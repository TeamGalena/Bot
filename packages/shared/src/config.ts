class ConfigurationError extends Error {}

export function requireEnv(key: string) {
  const value = import.meta.env[key] ?? process.env[key];
  if (value) return value;
  throw new ConfigurationError(`Missing environment variable '${key}'`);
}

export function optionalEnv(key: string) {
  try {
    return requireEnv(key);
  } catch {
    return null;
  }
}
