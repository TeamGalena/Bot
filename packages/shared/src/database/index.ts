export * from "./audit";
export { migrateDatabase, truncateDatabase } from "./connection";
export * from "./link";
export * from "./mod";
export * from "./role";
