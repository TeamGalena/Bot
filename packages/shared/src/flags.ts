import type { LinkEntry } from "./database";

const Flags = {
  pride: 0,
  anniversary: 1,
  gilded: 3,
  mehvahdjukaar: 2,
  tuccut: 3,
};

export type Flag = keyof typeof Flags;

export const SUPPORTER_FLAGS: Flag[] = ["pride"];

export const FLAGS = Object.keys(Flags) as Flag[];

export function createFlags(...values: Flag[]) {
  if (values.length === 0) return 0;
  const masks = values.map((it) => 1 << Flags[it]);
  return masks.reduce((a, b) => a | b);
}

export function withFlags(link: LinkEntry, ...added: Flag[]): LinkEntry {
  if (added.length === 0) return link;
  const flags = link.flags | createFlags(...added);
  return { ...link, flags };
}

export function isFlag(flag: string): flag is Flag {
  return flag in Flags;
}

export function flagQuery(flag: Flag) {
  if (!isFlag(flag)) throw new Error(`unknown flag ${flag}`);
  const mask = 1 << Flags[flag];
  return `(flags & ${mask} != 0)`;
}

export function extractFlags(mask: number): Flag[] {
  return Object.keys(Flags)
    .map((it) => it as Flag)
    .filter((flag) => {
      const flagMask = 1 << Flags[flag];
      return (flagMask & mask) !== 0;
    });
}
