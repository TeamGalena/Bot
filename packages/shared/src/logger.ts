/* eslint-disable no-console */

import chalk from "chalk";

export type Logger = {
  debug(message: string): void;
  info(message: string): void;
  warn(message: string): void;
  error(message: string, exception?: unknown): void;
};

const logger: Logger = {
  debug: (it) => console.log(chalk.blue(it)),
  info: (it) => console.log(it),
  warn: (it) => console.warn(chalk.yellow(it)),
  error: (it, ...exception) => console.error(chalk.red(it), ...exception),
};

export default logger;
