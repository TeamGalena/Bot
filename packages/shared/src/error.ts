import logger from "./logger";

export class UserError extends Error {}

export function wrapCatching<T extends unknown[]>(
  func: (...args: T) => void | Promise<void>
) {
  return async (...args: T) => {
    try {
      await func(...args);
    } catch (ex) {
      logger.error(`Encountered an error executing ${func.name}`, ex);
    }
  };
}

function wrapError<TArgs extends unknown[], TReturn>(
  func: (...args: TArgs) => Promise<TReturn>,
  resolver: (ex: Error) => Error
) {
  return async (...args: TArgs): Promise<TReturn> => {
    try {
      return await func(...args);
    } catch (ex) {
      if (ex instanceof Error) throw resolver(ex);
      else throw new Error(ex?.toString());
    }
  };
}

export function createErrorWrapper(resolver: (ex: Error) => Error) {
  return <TArgs extends unknown[], TReturn>(
    func: (...args: TArgs) => Promise<TReturn>
  ) => wrapError(func, resolver);
}
