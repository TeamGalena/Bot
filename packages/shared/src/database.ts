import { join } from "path";
import { open } from "sqlite";
import sqlite3 from "sqlite3";
import { optionalEnv } from "./config";
import { UserError } from "./error";
import { createFlags, flagQuery, withFlags, type Flag } from "./flags";
import logger from "./logger";
import type { Page, Paginated, Pagination } from "./paginated";
import type { User } from "./user";
import { repeat } from "./util";

const migrationsPath = optionalEnv("MIGRATIONS_PATH") ?? "/migrations";
const dataPath = optionalEnv("DATA_PATH") ?? "/data";

export async function migrateDatabase() {
  logger.debug(`searching for migrations in ${migrationsPath}`);
  await db.migrate({ migrationsPath });
  logger.info("migrated database");
}

logger.debug(`Loading database in ${dataPath}`);

const db = await open({
  filename: join(dataPath, "database.db"),
  driver: sqlite3.Database,
});

export type InputLinkEntry = {
  discordId: string;
  uuid: string;
  rank: number;
  flags?: number | Flag[];
};

export type LinkEntry = InputLinkEntry & {
  id: number;
  createdAt: string;
  updatedAt: string;
  flags: number;
};

export type RoleEntry = {
  id: string;
  rank: number;
};

function resolveError(ex: unknown) {
  if (!(ex instanceof Error)) return ex;
  if (!("code" in ex)) return ex;
  if (ex.code === "SQLITE_CONSTRAINT")
    return new UserError(
      "this minecraft uuid is already linked to another account"
    );

  return ex;
}

function wrapError<TArgs extends unknown[], TReturn>(
  func: (...args: TArgs) => Promise<TReturn>
) {
  return async (...args: TArgs): Promise<TReturn> => {
    try {
      return await func(...args);
    } catch (ex) {
      throw resolveError(ex);
    }
  };
}

export async function getLinkByDiscordId(discordId: string) {
  return await db.get<LinkEntry>("SELECT * FROM Link WHERE discordId = ?", [
    discordId,
  ]);
}

function encodeCursor(entry: LinkEntry) {
  return btoa(`link-${entry.id.toString()}`);
}

function decodeCursor(cursor?: string) {
  if (!cursor) return 0;
  const decoded = atob(cursor).substring("link-".length);
  const parsed = Number.parseInt(decoded);
  if (isNaN(parsed)) return 0;
  return parsed;
}

function createPage<T>(
  items: Paginated<T>[],
  size: number,
  total: number
): Page<T> {
  if (items.length > size) {
    return {
      items: items.slice(1),
      pageInfo: {
        size,
        total,
        next: items[items.length - 1].cursor,
      },
    };
  } else {
    return {
      items,
      pageInfo: {
        total,
        size: items.length,
      },
    };
  }
}

export type LinkFilter = {
  search?: string;
  flag?: Flag;
};

export async function getLinks(
  pagination: Pagination,
  filter: LinkFilter = {}
): Promise<Page<LinkEntry>> {
  const afterId = decodeCursor(pagination.after);

  const terms: string[] = ["TRUE"];
  const params: string[] = [];

  if (filter.search) {
    terms.push(`(uuid LIKE ?) OR (uuid LIKE ?)`);
    const pattern = `%${filter.search}%`;
    params.push(pattern);
    params.push(pattern);
  }
  if (filter.flag) terms.push(flagQuery(filter.flag));

  const query = terms.map((it) => `(${it})`).join(" AND ");
  logger.debug(query);

  const { total } = await db.get(
    `SELECT COUNT(*) total FROM Link WHERE ${query}`
  );

  const entries = await db.all<LinkEntry[]>(
    `SELECT * FROM Link WHERE id >= ? AND ${query} ORDER BY id ASC LIMIT ?`,
    afterId,
    ...params,
    pagination.size + 1
  );

  const items = entries.map((value) => ({
    value,
    cursor: encodeCursor(value),
  }));

  return createPage(items, pagination.size, total);
}

export async function getLinkByUuid(uuid: string) {
  return await db.get<LinkEntry>("SELECT * FROM Link WHERE uuid = ? LIMIT 1", [
    uuid,
  ]);
}

export const updateLink = wrapError(
  async (existing: LinkEntry, values: InputLinkEntry, user: User) => {
    const next = { ...existing, ...values };
    next.flags = resolveFlags(next.flags);
    if (
      next.rank === existing.rank &&
      next.uuid === existing.uuid &&
      next.flags === existing.flags
    ) {
      logger.debug("skipped linking, all values are the same");
      return;
    }

    await db.run(
      "UPDATE Link SET uuid = ?, rank = ?, flags = ?, updatedAt = datetime('now') WHERE discordId = ?",
      [next.uuid, next.rank, next.flags, next.discordId]
    );

    await addAudit({
      user,
      action: "update_link",
      subject: existing.discordId,
    });

    const prettyFlags = next.flags.toString(2).padStart(8, "0");
    logger.debug(
      `updated ${next.discordId} <-> ${next.uuid} (${next.rank}/${prettyFlags})`
    );
  }
);

function resolveFlags(flags: InputLinkEntry["flags"]): number {
  return typeof flags === "number" ? flags : createFlags(...(flags ?? []));
}

export const insertLink = wrapError(
  async (link: InputLinkEntry, user: User) => {
    await db.run(
      "INSERT INTO Link (discordId, uuid, rank, flags) VALUES (?, ?, ?, ?)",
      [link.discordId, link.uuid, link.rank, resolveFlags(link.flags)]
    );

    await addAudit({ user, action: "create_link", subject: link.discordId });
  }
);

export async function persistLink(link: InputLinkEntry, user: User) {
  const existing = await getLinkByDiscordId(link.discordId);

  if (existing) {
    await updateLink(existing, link, user);
  } else {
    await insertLink(link, user);

    logger.debug(`linked ${link.discordId} <-> ${link.uuid} (${link.rank})`);
  }
}

async function requireLink(discordId: string) {
  const link = await getLinkByDiscordId(discordId);
  if (link) return link;
  throw new UserError("you have not linked your minecraft account yet");
}

export async function addFlags(discordId: string, flags: Flag[], user: User) {
  const link = await requireLink(discordId);
  await updateLink(link, withFlags(link, ...flags), user);
}

export async function updateRank(discordId: string, rank: number, user: User) {
  const link = await requireLink(discordId);
  await updateLink(link, { ...link, rank }, user);
}

export async function loadSupporterRoles() {
  return await db.all<RoleEntry[]>("SELECT * FROM Role ORDER BY rank DESC");
}

export async function containsAdminRole(roles: string[]) {
  const { count } = await db.get(
    `SELECT COUNT(*) count FROM Role WHERE isAdmin = TRUE AND id IN (${repeat(
      "?",
      roles.length
    )})`,
    ...roles
  );

  return count > 0;
}

export async function deleteLinkByDiscordId(discordId: string, user: User) {
  await db.run(`DELETE FROM Link WHERE discordId = ?`, [discordId]);
  await addAudit({ user, action: "delete_link", subject: discordId });
}

export async function truncateLinks() {
  await db.run(`DELETE FROM Link WHERE 1`);
}

type InputAuditEntry = {
  user: User;
  action: string;
  subject: string;
};

type AuditEntry = InputAuditEntry & {
  date: string;
};

async function addAudit({ action, user, subject }: InputAuditEntry) {
  await db.run(
    `INSERT INTO AuditLog (user, action, subject) VALUES (?, ?, ?)`,
    [user.username, action, subject]
  );
}

export async function getAuditLog() {
  return await db.all<AuditEntry[]>(
    "SELECT * FROM AuditLog ORDER BY date DESC LIMIT 100"
  );
}
