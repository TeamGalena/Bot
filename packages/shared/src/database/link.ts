import { createErrorWrapper, UserError } from "../error";
import { createFlags, flagQuery, withFlags, type Flag } from "../flags";
import logger from "../logger";
import type { Page, Pagination } from "../paginated";
import type { User } from "../user";
import { now } from "../util";
import { addAudit } from "./audit";
import { db } from "./connection";
import { createPage } from "./paginated";

const wrapError = createErrorWrapper((ex) => {
  if (!("code" in ex)) return ex;
  if (ex.code === "SQLITE_CONSTRAINT")
    return new UserError(
      "this minecraft uuid is already linked to another account"
    );

  return ex;
});

export type InputLinkEntry = {
  discordId: string;
  uuid: string;
  rank: number;
  flags?: number | Flag[] | Flag;
};

export type LinkEntry = InputLinkEntry & {
  id: number;
  createdAt: string;
  updatedAt: string;
  flags: number;
};

export type LinkFilter = {
  search?: string;
  flag?: Flag;
};

export async function getLinkByDiscordId(discordId: string) {
  return await db.get<LinkEntry>("SELECT * FROM Link WHERE discordId = ?", [
    discordId,
  ]);
}

function encodeCursor(entry: LinkEntry) {
  return btoa(`link-${entry.id.toString()}`);
}

function decodeCursor(cursor?: string) {
  if (!cursor) return 0;
  const decoded = atob(cursor).substring("link-".length);
  const parsed = Number.parseInt(decoded);
  if (isNaN(parsed)) return 0;
  return parsed;
}

export async function getLinks(
  pagination: Pagination,
  filter: LinkFilter = {}
): Promise<Page<LinkEntry>> {
  const afterId = decodeCursor(pagination.after);

  const terms: string[] = ["TRUE"];
  const params: string[] = [];

  if (filter.search) {
    terms.push(`(uuid LIKE ?) OR (uuid LIKE ?)`);
    const pattern = `%${filter.search}%`;
    params.push(pattern);
    params.push(pattern);
  }
  if (filter.flag) terms.push(flagQuery(filter.flag));

  const query = terms.map((it) => `(${it})`).join(" AND ");

  const { total } = await db.get(
    `SELECT COUNT(*) total FROM Link WHERE ${query}`
  );

  const entries = await db.all<LinkEntry[]>(
    `SELECT * FROM Link WHERE id >= ? AND ${query} ORDER BY id ASC LIMIT ?`,
    afterId,
    ...params,
    pagination.size + 1
  );

  const items = entries.map((value) => ({
    value,
    cursor: encodeCursor(value),
  }));

  return createPage(items, pagination.size, total);
}

export async function getLinkByUuid(uuid: string) {
  return await db.get<LinkEntry>("SELECT * FROM Link WHERE uuid = ? LIMIT 1", [
    uuid,
  ]);
}

export const updateLink = wrapError(
  async (existing: LinkEntry, values: InputLinkEntry, user: User) => {
    const next = { ...existing, ...values };
    next.flags = resolveFlags(next.flags);
    if (
      next.rank === existing.rank &&
      next.uuid === existing.uuid &&
      next.flags === existing.flags
    ) {
      logger.debug("skipped linking, all values are the same");
      return;
    }

    await db.run(
      "UPDATE Link SET uuid = ?, rank = ?, flags = ?, updatedAt = ? WHERE discordId = ?",
      [next.uuid, next.rank, next.flags, now(), next.discordId]
    );

    await addAudit({
      user,
      action: "update_link",
      subject: existing.discordId,
    });

    const prettyFlags = next.flags.toString(2).padStart(8, "0");
    logger.debug(
      `updated ${next.discordId} <-> ${next.uuid} (${next.rank}/${prettyFlags})`
    );
  }
);

function resolveFlags(flags: InputLinkEntry["flags"]): number {
  if (typeof flags === "number") return flags;
  if (Array.isArray(flags)) return createFlags(...flags);
  if (!flags) return createFlags();
  return createFlags(flags);
}

export const insertLink = wrapError(
  async (link: InputLinkEntry, user: User) => {
    await db.run(
      "INSERT INTO Link (discordId, uuid, rank, flags) VALUES (?, ?, ?, ?)",
      [link.discordId, link.uuid, link.rank, resolveFlags(link.flags)]
    );

    await addAudit({ user, action: "create_link", subject: link.discordId });
  }
);

export async function persistLink(link: InputLinkEntry, user: User) {
  const existing = await getLinkByDiscordId(link.discordId);

  if (existing) {
    await updateLink(existing, link, user);
  } else {
    await insertLink(link, user);

    logger.debug(`linked ${link.discordId} <-> ${link.uuid} (${link.rank})`);
  }
}

async function requireLink(discordId: string) {
  const link = await getLinkByDiscordId(discordId);
  if (link) return link;
  throw new UserError("you have not linked your minecraft account yet");
}

export async function addFlags(discordId: string, flags: Flag[], user: User) {
  const link = await requireLink(discordId);
  await updateLink(link, withFlags(link, ...flags), user);
}

export async function updateRank(discordId: string, rank: number, user: User) {
  const link = await requireLink(discordId);
  await updateLink(link, { ...link, rank }, user);
}

export async function deleteLinkByDiscordId(discordId: string, user: User) {
  await db.run(`DELETE FROM Link WHERE discordId = ?`, [discordId]);
  await addAudit({ user, action: "delete_link", subject: discordId });
}
