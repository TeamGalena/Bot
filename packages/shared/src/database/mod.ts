import { UserError } from "../error";
import { db } from "./connection";

export type ModEntry = {
  name: string;
  icon?: string;
  color: `#${string}`;
  description?: string;
  repository: string;
  curseforgeSlug?: string;
  modrinthSlug?: string;
};

export async function getModBySearch(search: string) {
  const pattern = `%${search}%`;
  const mod = await db.get<ModEntry>(
    "SELECT * FROM ModInfo WHERE id LIKE ? OR name LIKE ? COLLATE NOCASE",
    [pattern, pattern]
  );
  if (mod) return mod;
  throw new UserError(`Cannot find any mod named '${search}'`);
}

export async function getMods() {
  return await db.all<ModEntry[]>(
    "SELECT * FROM ModInfo ORDER BY name LIMIT 20"
  );
}
