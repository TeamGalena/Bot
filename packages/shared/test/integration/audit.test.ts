import { afterEach, beforeAll, expect, setSystemTime, test } from "bun:test";
import {
  deleteLinkByDiscordId,
  getAuditLog,
  getLinkByDiscordId,
  insertLink,
  LinkEntry,
  migrateDatabase,
  truncateDatabase,
  updateLink,
} from "../../src/database";
import { User } from "../../src/user";
import { createTestLinkInput } from "../helper/link";

const user: User = {
  username: "test-user",
};

beforeAll(async () => {
  await migrateDatabase();
});

afterEach(async () => {
  await truncateDatabase();
  setSystemTime();
});

test("creates audit log when creating an entry", async () => {
  setSystemTime(new Date("2025-02-12 14:33:47"));

  const link = createTestLinkInput();
  await insertLink(link, user);

  const log = await getAuditLog();

  expect(log).toHaveLength(1);
  expect(log[0]).toMatchSnapshot("insert-audit-log");
});

test("creates audit log when updating an entry", async () => {
  setSystemTime(new Date("2025-02-12 14:33:47"));

  const link = createTestLinkInput();
  const update = createTestLinkInput({
    uuid: "updated-test-uuid",
    flags: ["anniversary", "gilded"],
  });

  // This should be some form of fixture, might be possible when using an ORM
  await insertLink(link, user);
  const created = await getLinkByDiscordId(link.discordId);

  setSystemTime(new Date("2025-02-16 14:33:47"));
  await updateLink(created as LinkEntry, update, user);

  const log = await getAuditLog();

  expect(log).toHaveLength(2);
  expect(log[0]).toMatchSnapshot("update-audit-log");
});

test("creates audit log when deleting an entry", async () => {
  setSystemTime(new Date("2025-02-12 14:33:47"));

  const link = createTestLinkInput();

  // This should be some form of fixture, might be possible when using an ORM
  await insertLink(link, user);

  setSystemTime(new Date("2025-02-16 14:33:47"));
  await deleteLinkByDiscordId(link.discordId, user);

  const log = await getAuditLog();

  expect(log).toHaveLength(2);
  expect(log[0]).toMatchSnapshot("delete-audit-log");
});
