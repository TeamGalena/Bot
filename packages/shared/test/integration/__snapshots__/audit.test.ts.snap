// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`creates audit log when creating an entry: insert-audit-log 1`] = `
{
  "action": "create_link",
  "date": "2025-02-12T14:33:47.000Z",
  "id": 1,
  "subject": "test-discord-id",
  "user": "test-user",
}
`;

exports[`creates audit log when updating an entry: update-audit-log 1`] = `
{
  "action": "update_link",
  "date": "2025-02-16T14:33:47.000Z",
  "id": 2,
  "subject": "test-discord-id",
  "user": "test-user",
}
`;

exports[`creates audit log when deleting an entry: delete-audit-log 1`] = `
{
  "action": "delete_link",
  "date": "2025-02-16T14:33:47.000Z",
  "id": 2,
  "subject": "test-discord-id",
  "user": "test-user",
}
`;
